aten/src/ATen/native/GridSampler.cpp

```diff
@@ -380,12 +381,278 @@ namespace {
+                  scalar_t nw_val = inp_ptr_NC[iy_nw * inp_sH + ix_nw * inp_sW];
+                  gix -= nw_val * (iy_se - iy) * gOut;
+                  giy -= nw_val * (ix_se - ix) * gOut;
+                }
+                if (within_bounds_2d(iy_ne, ix_ne, inp_H, inp_W)) {
+                  scalar_t ne_val = inp_ptr_NC[iy_ne * inp_sH + ix_ne * inp_sW];
+                  gix += ne_val * (iy_sw - iy) * gOut;
+                  giy -= ne_val * (ix - ix_sw) * gOut;
+                }
+                if (within_bounds_2d(iy_sw, ix_sw, inp_H, inp_W)) {
+                  scalar_t sw_val = inp_ptr_NC[iy_sw * inp_sH + ix_sw * inp_sW];
+                  gix -= sw_val * (iy - iy_ne) * gOut;
+                  giy += sw_val * (ix_ne - ix) * gOut;
+                }
+                if (within_bounds_2d(iy_se, ix_se, inp_H, inp_W)) {
+                  scalar_t se_val = inp_ptr_NC[iy_se * inp_sH + ix_se * inp_sW];
+                  gix += se_val * (iy - iy_nw) * gOut;
+                  giy += se_val * (ix - ix_nw) * gOut;
+                }
+              }
+
+              // assuming grad_grid is contiguous
+              gGrid_ptr_NHW[0] = gix_mult * gix;
+              gGrid_ptr_NHW[1] = giy_mult * giy;
+            } else if (interpolation_mode == GridSamplerInterpolation::Nearest) {
+              int64_t ix_nearest = static_cast<int64_t>(std::round(ix));
+              int64_t iy_nearest = static_cast<int64_t>(std::round(iy));
+
+              // assign nearest neighor pixel value to output pixel
+              scalar_t *gOut_ptr_NCHW = gOut_ptr + n * gOut_sN + h * gOut_sH + w * gOut_sW;
+              scalar_t *gInp_ptr_NC = gInp_ptr + n * gInp_sN;
+              for (int c = 0; c < C; ++c, gOut_ptr_NCHW += gOut_sC, gInp_ptr_NC += gInp_sC) {
+                // calculate and set grad_input
+                safe_add_2d(gInp_ptr_NC, iy_nearest, ix_nearest, gInp_sH, gInp_sW,
+                            inp_H, inp_W, *gOut_ptr_NCHW);
+              }
+            }
+          }
+        }
+      }
+    });
+    return std::make_tuple(grad_input, grad_grid);
+  }
+
 }  // namespace
 
 // No shape checking needed here. See # NOTE [ grid_sampler Native Functions ].
 Tensor grid_sampler_2d_cpu(const Tensor& input, const Tensor& grid,
                            int64_t interpolation_mode, int64_t padding_mode,
                            bool align_corners) {
+  // AVX gather instructions use signed 32-bit offsets to gather float values.
+  // Check for possible overflow and fallback to scalar implementation
+  if (input.scalar_type() != kDouble) {
+    TORCH_CHECK(input.scalar_type() == kFloat,
+                "grid_sampler_2d_cpu not implemented for ", input.scalar_type());
+    auto sizes = input.sizes();
+    auto strides = input.strides();
+    // NOTE: Gather offsets are only used for the height and width dimensions
+    auto max_gather_offset =
+      (sizes[2] - 1) * strides[2] + (sizes[3] - 1) * strides[3];
+
```

Comment:
If I am reading this correctly, https://github.com/pytorch/pytorch/blob/5ff54ff4ff0dadbdbf909addf3180d13cc16679a/aten/src/ATen/native/cpu/GridSamplerKernel.cpp#L875-L877, it looks like we do a gather on the grid. The gather instruction has the problem where if the input is a float, then it'll use an int32 index. So should we also check the gather offset for the grid (in both the forward & backward pass)?

Reply:
I saw the `arange(0, grid_sW)` above and assumed this would be contiguous like python `range`. However, it's actually a stride of `sW` so I've added a check for that. The code was also increasing the vector `i_offsets` by a constant vector which I changed to just bumping the base ptr for the gather instead, that way we are even less unlikely to run out of 32-bits.